qmi/core/transport.py

```diff
@@ -858,70 +853,105 @@ def read(self, nbytes: int, timeout: Optional[float]) -> bytes:
+                raise QMI_EndOfInputException(
+                    f"The read buffer did not contain expected bytes of data ({nbuf} != {nbytes}."
+                )
 
     def read_until(self, message_terminator: bytes, timeout: Optional[float]) -> bytes:
-        """Read a single USBTMC message from the instrument."
+        """The specified message_terminator is ignored. Instead, the instrument must autonomously indicate the end
+        of the message according to the USBTMC protocol.
 
-        The specified message_terminator is ignored.
-        Instead, the instrument must autonomously indicate the end of the
-        message according to the USBTMC protocol.
+        As the message_terminator is not used, we forward simply the call to the `read_until_timeout` call.
 
-        This method blocks until a complete USBTMC message is available.
+        Parameters:
+            message_terminator: This input is ignored.
+            timeout: Maximum time to wait (in seconds).
+
+        Returns:
+            Received bytes.
+        """
+        return self.read_until_timeout(0, timeout)
+
+    def read_until_timeout(self, nbytes: int, timeout: float) -> bytes:
+        """Read a single USBTMC message from the instrument.
+
+        If there is already data in the buffer, return it.
 
         If the timeout expires before the message is received, the read is
         aborted and any data already received are discarded. In this
         case QMI_TimeoutException is raised.
 
         Parameters:
-            message_terminator: Message terminator characters.
+            nbytes: This input is ignored.
             timeout: Maximum time to wait (in seconds).
 
         Returns:
-            Received bytes, including the terminator.
+            Received bytes.
 
         Raises:
             ~qmi.core.exceptions.QMI_TimeoutException: If the timeout expires before the
-                requested number of bytes are available.
-
+            requested number of bytes are available.
         """
         self._check_is_open()
 
         # USB requires a timeout
         if timeout is None:
             timeout = self.DEFAULT_READ_TIMEOUT
 
+        data = bytes()
         if self._read_buffer:
             # Data already available in buffer; return it now.
-            data = self._read_buffer
+            data += self._read_buffer
             self._read_buffer = bytes()
+
         else:
-            # Read buffer is empty - read a new message from the instrument.
-            data = self._read_message(timeout)
+            # Read a new message from the instrument.
+            data += self._read_message(timeout)
 
         return data
 
+    def discard_read(self) -> None:
+        # We should empty the buffer, or if it is empty, discard the next message from the source
+        if not len(self._read_buffer):
+            # Read buffer is empty - read a new message from the instrument.
+            try:
+                self._read_message(0.0)
```

Comment:
what happens if there's multiple message in the queue? don't you want to read until timeout?

Reply:
USB should not have multiple messages in the queue, but anyhow if the user wants to discard multiple reads, then is should be called as many times instead of clearing everything. I think this is more the idea of this function.